src/components/UploadButton/UploadButton.tsx

```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import './UploadButton.css';
+
+interface PropTypes {
+  // Button contents
+  label?: string;
+  // This type comes from a built-in React method.
+  // Essentially, its typed to the "setState" function.
+  setFile: React.Dispatch<React.SetStateAction<string | null>>;
+};
+
+// Primary UI component for user interaction
+const UploadButton = ({
+  label="Upload a File...",
+  setFile,
+  ...props
+}: PropTypes) => {
+  // Set the state of the file.
+  return (
+    <div>
+      <label htmlFor="file-uploader">
+        {label}
+      </label>
+
+      <input
+        id="file-uploader"
+        name="file"
+        type="file"
+        onChange={(onChangeEvent) => {
+          // Get the data of this file uploader HTML element.
+          const fileUploader = onChangeEvent.currentTarget
+          // Make sure it has the files.
+          if (fileUploader.files === null || fileUploader.files.length < 1) {
+            return
+          }
+
+          // Per the documentation of <input>, there is only one file
+          // in the files array unless the multiple property is enabled.
+          // Also, file is type "File", which is a JS Web API.
+          const selectedFile: File = fileUploader.files[0]
+
+          // Use the "FileReader" JS Web API to read variables with type "File".
+          const reader: FileReader = new FileReader()
```

Comment:
I would move the file reading logic somewhere else. It might be useful to have the full `File` object (for example to grab the file name - or to do a streaming parser).

Here is a little exercise for you, can you write the file reader logic as a hook?

It would work something like this:
```ts
  const [file, setfile] = useState<File>()
  const fileContent = useFileContentAsString(file)

  return (
    <UploadButton setFile={setFile} />
    <textbox value={fileContent} />
  )
```

Reply:
I've set up a custom hook that creates an object in state in recent commits (see https://github.com/AsyncRacing/AsyncRacing/pull/5/commits/a5b364f88ada64bad24be2590e3bc6a140625e41and earlier). It takes in a file, reads it, and spits out an object containing the file, the file's text, and potentially other things in the future.